### ISSUE TITLE
Async dialog API is slow to return on Linux

### ISSUE THREAD

**Issue by oNaiPs** [electron/electron#20533]
### Preflight Checklist
* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details
When using async dialog API, the function takes several seconds to return.
This can be observed in `showSaveDialog`, `showMessageBox` and `showOpenDialog`.

* **Electron Version:**
v7.0.0-beta.6
v6.0.12
* **Operating System:**
Ubuntu 19.04 x64
* **Last Known Working Electron version:**
v5.0.11

### Expected Behavior
There shouldn't be any added delay after user clicks on the dialog.

### Actual Behavior
API only returns after several seconds after user clicked on a dialog option.

### To Reproduce

Run the example code:
```
const { app, dialog } = require('electron');

app.on('ready', () => {
  (async () => {
    console.log('Before Dialog!');
    const res = await dialog.showMessageBox({
      buttons: [
        'ok',
      ],
      title: 'Title',
      message: 'This is a message.',
    });

    console.log('After Dialog');
    console.log(`Res: ${ JSON.stringify(res) }`);
  })().catch(e => { console.error(e); });
});
```
Observe the delay after clicking on "OK".

### Screenshots
Bug reproduction screencast:
https://youtu.be/ZfiobQDvbOs

**Comment by Liberation927** [comment]
I came to find this problem on github, and it really has

**Comment by lorenzogrv** [comment]
After I noticed this today, and found this issue, I would like to confirm this bug. Note the synchronous version returns way faster than asynchronous one:

Test script:

```
$ cat test-slow-dialog.js 
const { app, dialog } = require('electron');

app.on('ready', () => {
  console.log('Before Dialog!');
  let box = {
    buttons: [ 'ok' ],
    title: 'Title',
    message: 'This is a message.'
  }
  let now = Date.now()
  dialog
    .showMessageBox(box)
    .then(res => {
      console.log('After promised dialog (%s ms)', Date.now() - now)
      console.log(`Res: ${ JSON.stringify(res) }`)
      return Date.now()
    })
    .then(now => {
      let res = dialog.showMessageBoxSync(box)
      console.log('After synchronous dialog (%s ms)', Date.now() - now)
      console.log(`Res: ${ JSON.stringify(res) }`)
    })
    .catch(e => console.error(e))
    .finally(() => app.quit())
});
```

Running the test:

```
$ electron test-slow-dialog.js 
Before Dialog!
After promised dialog (8041 ms)
Res: {"response":0,"checkboxChecked":false}
After synchronous dialog (806 ms)
Res: 0
```

Env:

```
$ electron --version
v8.0.2
$ uname -srv
Linux 5.5.6-genuine #1 SMP Wed Feb 26 03:38:43 CET 2020
```

**Comment by m4heshd** [comment]
Can confirm this issue on Windows too. Electron v12.0.0. It takes about 3 seconds for the promise to be returned. Sync version has no issue.

**Comment by nornagon** [comment]
I can't reproduce this on Windows under Electron v12.0.0:

```
Before Dialog!
After promised dialog (413 ms)
Res: {"response":0,"checkboxChecked":false}
After synchronous dialog (209 ms)
Res: 0
```

**Comment by nornagon** [comment]
I can reproduce it on Linux, though not consistently:

```
Before Dialog!
Gtk-Message: 15:37:22.921: GtkDialog mapped without a transient parent. This is discouraged.
libva error: va_getDriverName() failed with unknown libva error,driver_name=(null)
After promised dialog (4994 ms)
Res: {"response":0,"checkboxChecked":false}
Gtk-Message: 15:37:27.902: GtkDialog mapped without a transient parent. This is discouraged.
After synchronous dialog (821 ms)
Res: 0
```

having a `setInterval` running in the background "patches over" this issue, which makes me think it's an event-loop integration problem.

workaround:

```js
const { app, dialog } = require('electron');

app.on('ready', () => {
  console.log('Before Dialog!');
  let interval = setInterval(() => { /* nothing */ }, 100)
  let box = {
    buttons: [ 'ok' ],
    title: 'Title',
    message: 'This is a message.'
  }
  let now = Date.now()
  dialog
    .showMessageBox(box)
    .then(res => {
      clearInterval(interval)
      console.log('After promised dialog (%s ms)', Date.now() - now)
      console.log(`Res: ${ JSON.stringify(res) }`)
      return Date.now()
    })
    .then(now => {
      let res = dialog.showMessageBoxSync(box)
      console.log('After synchronous dialog (%s ms)', Date.now() - now)
      console.log(`Res: ${ JSON.stringify(res) }`)
    })
    .catch(e => console.error(e))
    .finally(() => app.quit())
});
```

**Comment by m4heshd** [comment]
@nornagon as you said It's not consistent on Windows either. Works fine after a reboot most of the time. What I noticed is it's something related to Windows explorer because explorer sometimes crash or act buggy when I try this. Also the issue is only present for me on open/save dialogs. Probably caused by a Windows update. Async dialog API runs on a separate thread right? So under the hood it's still sync?

**Comment by lorenzogrv** [comment]
@nornagon I suspect your conclusion - event loop problem - is right. I would test also something like

```js
process.setImmediate(() => {
  dialog
    .showMessageBox(box)
    // ....
})
```

to research even more on the subject

**Comment by Uninen** [comment]
I bumped into a related issue today. I'm inexperienced with Electron so it took me quite a while to figure out what's going on. But my problem went away after applying the interval workaround so I'm assuming it's related.

I'm calling `dialog.showOpenDialog` in main process from the render process via a `contextBridge` api. In the handler I'm resolving the dialog promise and then sending `result.filePaths` back to the render process via another `contextBridge` api. This all works great -- but only once or twice.

After one or two repeated dialog opens, the render process just hangs. There's no error messages in the console and the devtools keep working, but the window freezes. The last time the dialog closes, the results are sent okay but they never arrive and the render window freezes.

After finding this ticket, I tried adding no-op setInterval to the main process handler like in the workaround above -- and now my code works as it should.

I'm running the latest Electron (17) on macOS 12.2.1.
